### PR TITLE
Auto tray reorganization

### PR DESCRIPTION
**Pull request by xzinkus**
Alright, I believe I implemented everything we discussed. Can you test this on arch with normal `netctl`? It works on Ubuntu with `netctl-auto` and will still work on Arch for `-auto`.

Please see commit message of the last commit for more detailed change docs.

### CODE REVIEW

**Comment by xzinkus** [comment]
Closes #2

**Comment by xzinkus** [comment]
If you want `netctl-auto`, compile with `cargo build --release --features auto`, otherwise leave off the `--features auto`.

**Comment by xzinkus** [comment]
I'm not sure what's needed to make the AUR able to compile with `--features auto` if the user decides, so I cleared out the installation instructions. Once we merge this, I figure you can add back the AUR steps in addition to leaving `install.sh` there for non-Arch users.

Thanks!

File: src/main.rs [hunk 1]
```diff
@@ -131,8 +131,8 @@ fn main() {
 				);
 			});
 			let mut update_timer = QTimer::new_0a();
-			// Call it every second
-			update_timer.set_interval(1000);
+			// Call it every 2 seconds
```
Comment (PonasKovas) [inline src/main.rs:122]: i guess this depends on the user, we should make this configurable

**Review by PonasKovas — CHANGES_REQUESTED** [review]
> commit 6ffc4ecf4bf05153d57dea22d0a65a9e46458dd4

I'd prefer that we kept this out of the repository, you can have it in local storage if you need it

File: Cargo.toml [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
 qt_widgets = "0.3.0"
 qt_gui = "0.3.0"
 regex = "1.3.1"
-shells = "0.2.0"
\ No newline at end of file
+shells = "0.2.0"
+
+[[bin]]
+name = "netctl-tray"
+path = "src/main.rs"
```
Comment (PonasKovas) [inline Cargo.toml:15]: what's the purpose of this?

File: README.md [hunk 3]
```diff
@@ -11,19 +11,21 @@ To launch the tray app:
...
-$ cd netctl-tray
-$ makepkg -si
+$ cargo build --release # --features auto if netctl-auto is used
+$ sudo ./install
 ```
 
 ## Troubleshooting
 
+If you have a `netctl` profile named `None` you may run into issues.
+
```
Comment (PonasKovas) [inline README.md:28]: how come? what causes it?

File: scripts/connect [hunk 4]
```diff
@@ -3,6 +3,5 @@
 display=":$(ls /tmp/.X11-unix/* | sed 's#/tmp/.X11-unix/X##' | head -n 1)"
 user=$(who | grep '('$display')' | awk '{print $1}' | head -n 1)
 uid=$(id -u $user)
-sleep 0.1
```
Comment (PonasKovas) [inline scripts/connect:6]: why did you remove this?

File: scripts/netctl-tray.policy [hunk 5]
```diff
@@ -0,0 +1,48 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (PonasKovas) [inline scripts/netctl-tray.policy:1]: what is the purpose of this file?

File: src/main.rs [hunk 6]
```diff
@@ -28,25 +12,26 @@ fn main() {
 		Err(_)=> false,
 	};
 	if as_root {
-		println!("Warning: tray started as root! This is unsafe!");
+		println!("Warning: tray started as root! Aborting.");
+      return;
```
Comment (PonasKovas) [inline src/main.rs:16]: why? its the user's choice

File: src/main.rs (hunk 1 above)
Comment (xzinkus) [reply src/main.rs:122]: Fair enough, I'll make it a command-line option.

File: README.md (hunk 3 above)
Comment (xzinkus) [reply README.md:28]: This was true before. Your code set the active profile name to "none" and then iterated through the profiles to find the active one, then checked to see if the name was still "none". I changed it to "None" so that it would look nicer in notifications etc, and documented the issue. I can clean up the code with a variable instead of the original sentinel value design.

File: scripts/connect (hunk 4 above)
Comment (xzinkus) [reply scripts/connect:6]: It seems to work without it, and sleeping is sort of an anti-pattern. If something needs to be awaited, then a synchronous call or checking a variable/value should be used.

File: scripts/netctl-tray.policy (hunk 5 above)
Comment (xzinkus) [reply scripts/netctl-tray.policy:1]: This is for Polkit version <0.106. >=0.106 allows use of the javascript .rules files, but 0.105 and earlier don't. In order to allow the pkexec calls, I needed to make them explicit actions. And then rather than add them as both actions with default deny and a .pkla polkit localauthority file to allow them, I just added them as actions with default allow. Which is fine, because they're minimal scripts, without real attack surface. One lets you call netctl list, and the other lets you call switch-to but only to a profile that already exists.

I guess an attacker who had local code execution could get you to disconnect from your current profile, but they could already do that without local code execution by de-authing your wifi. I think the risk is pretty low.

Comment (xzinkus) [reply scripts/netctl-tray.policy:1]: current Ubuntu ships polkit 0.105

File: src/main.rs (hunk 6 above)
Comment (xzinkus) [reply src/main.rs:16]: I was never clear on why it's dangerous to run this as root, but if it is, why not just disallow it?

**Comment by xzinkus** [comment]
> > commit [6ffc4ec](https://github.com/PonasKovas/netctl-tray/commit/6ffc4ecf4bf05153d57dea22d0a65a9e46458dd4)
> 
> I'd prefer that we kept this out of the repository, you can have it in local storage if you need it

This is super useful for anyone not on Arch, as they don't have to figure it out the way I did. Can't it stay in, with a comment in README that the preferred installation is via AUR?

**Comment by xzinkus** [comment]
TODO @maxzinkus:
- CLI option for refresh rate
- "none" profile name sentinel value refactor
- Remove `[[bin]]` dict from cargo.toml

File: Cargo.toml (hunk 2 above)
Comment (xzinkus) [reply Cargo.toml:15]: This might be redundant now, I can remove it.

File: src/main.rs (hunk 6 above)
Comment (PonasKovas) [reply src/main.rs:16]: > https://wiki.archlinux.org/index.php/Running_GUI_applications_as_root

just because it's dangerous we shouldn't forbid it completely, maybe the user knows what he's doing and really needs to run it as root

**Comment by PonasKovas** [comment]
I don't really understand why you're using `pkexec` and all those scripts?
Why didn't you just extend the polkit rule as we discussed earlier?

Also, after briefly skimming through the codebase I noticed lots of spaghetti, inefficient code and stuff like that and I'd like to completely refactor everything myself, as I don't have much to do during the quarantine. So I am not going to accept this PR, sorry.

And please, next time try to keep your different changes in separate PRs, it's really hard to focus when there are so many unrelated changes in a PR.

File: scripts/connect (hunk 4 above)
Comment (PonasKovas) [reply scripts/connect:6]: It didn't work without it for me, that's why I added it. Netctl calls those scripts before it has actually updated the profile data so we need to sleep a bit so we can say which profile was started

**Comment by xzinkus** [comment]
Alright. I'll keep maintaining my fork for now and merge your future work over mine. `pkexec` is the replacement for `sudo` when using PolKit, and the scripts are a bit messier but using them restricts the polkit rules to only allowing those specific actions. This is helpful because netctl-auto abstracts the systemd service a bit more than normal netctl.

Thanks for the quick responses and continuing to maintain!

**Comment by PonasKovas** [comment]
No, thank you for actively contributing!

File: scripts/netctl-tray.policy (hunk 5 above)
Comment (PonasKovas) [reply scripts/netctl-tray.policy:1]: so this file is for compatibility with older polkit versions? Was it generated? If so, how did you generate it? Its pretty unreadable

Comment (xzinkus) [reply scripts/netctl-tray.policy:1]: I wrote it based on the example for pkexec. All the action files in that directory look like this.
